<a href="https://colab.research.google.com/github/jdmartinev/MachineLearning/blob/main/Clase01/Clase01a_Food_delivery.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# John's food delivery

En este notebook, construiremos un clasificador de $k$ vecinos más cercanos para predecir el si las entregan llegarán a tiempo.

1. <a href="#1">Base de datos</a>
2. <a href="#2">Selección de características para entrenar el modelo</a>
3. <a href="#3">Entrenamiento del clasificador</a> ([__$K$ Nearest Neighbors Classifier__](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html))
4. <a href="#4">Uso del clasificador entrenado para predecir</a>
5. <a href="#5">Evaluación del modelo</a>
6. <a href="#6">Conjuntos de entrenamiento y prueba</a>
7. <a href="#7">Sobre-entrenamiento</a>



## 1. <a name="1">Base de datos</a>

Para este ejercicio, vamos a crear la base de datos


In [ ]:
import numpy as np

In [ ]:
data = np.array([[0, 1, 5, 1, 0],  # record of John's 1st food delivery
                [1, 0, 7, 0, 1],   # record of John's 2nd food delivery
                [0, 1, 2, 1, 0],   # record of John's 3rd food delivery
                [1, 1, 4.2, 1, 0], # record of John's 4th food delivery
                [0, 0, 7.8, 0, 1], # ...
                [1, 0, 3.9, 1, 0],
                [0, 1, 4, 1, 0],
                [1, 1, 2, 0, 0],
                [0, 0, 3.5, 0, 1],
                [1, 0, 2.6, 1, 0],
                [0, 0, 4.1, 0, 1],
                [0, 1, 1.5, 0, 1],
                [1, 1, 1.75, 1, 0],
                [1, 0, 1.3, 0, 0],
                [1, 1, 2.1, 0, 0],
                [1, 1, 0.2, 1, 0],
                [1, 1, 5.2, 0, 1],
                [0, 1, 2, 1, 0],
                [1, 0, 5.5, 0, 1],
                [0, 0, 2, 1, 0],
                [1, 1, 1.7, 0, 0],
                [0, 1, 3, 1, 1],
                [1, 1, 1.9, 1, 0],
                [0, 1, 3.1, 0, 1],
                [0, 1, 2.3, 0, 0],
                [0, 0, 1.1, 1, 0],
                [1, 1, 2.5, 1, 1],
                [1, 1, 5, 0, 1],
                [1, 0, 7.5, 1, 1],
                [0, 0, 0.5, 1, 0],
                [0, 0, 1.5, 1, 0],
                [1, 0, 3.2, 1, 0],
                [0, 0, 2.15, 1, 0],
                [1, 1, 4.2, 0, 1],
                [1, 0, 6.5, 0, 1],
                [1, 0, 0.5, 0, 0],
                [0, 0, 3.5, 0, 1],
                [0, 0, 1.75, 0, 0],
                [1, 1, 5, 0, 1],
                [0, 0, 2, 1, 0],
                [0, 1, 1.3, 1, 1],
                [0, 1, 0.2, 0, 0],
                [1, 1, 2.2, 0, 0],
                [0, 1, 1.2, 1, 0],
                [1, 1, 4.2, 0, 1]])

print(data)

[[0.   1.   5.   1.   0.  ]
 [1.   0.   7.   0.   1.  ]
 [0.   1.   2.   1.   0.  ]
 [1.   1.   4.2  1.   0.  ]
 [0.   0.   7.8  0.   1.  ]
 [1.   0.   3.9  1.   0.  ]
 [0.   1.   4.   1.   0.  ]
 [1.   1.   2.   0.   0.  ]
 [0.   0.   3.5  0.   1.  ]
 [1.   0.   2.6  1.   0.  ]
 [0.   0.   4.1  0.   1.  ]
 [0.   1.   1.5  0.   1.  ]
 [1.   1.   1.75 1.   0.  ]
 [1.   0.   1.3  0.   0.  ]
 [1.   1.   2.1  0.   0.  ]
 [1.   1.   0.2  1.   0.  ]
 [1.   1.   5.2  0.   1.  ]
 [0.   1.   2.   1.   0.  ]
 [1.   0.   5.5  0.   1.  ]
 [0.   0.   2.   1.   0.  ]
 [1.   1.   1.7  0.   0.  ]
 [0.   1.   3.   1.   1.  ]
 [1.   1.   1.9  1.   0.  ]
 [0.   1.   3.1  0.   1.  ]
 [0.   1.   2.3  0.   0.  ]
 [0.   0.   1.1  1.   0.  ]
 [1.   1.   2.5  1.   1.  ]
 [1.   1.   5.   0.   1.  ]
 [1.   0.   7.5  1.   1.  ]
 [0.   0.   0.5  1.   0.  ]
 [0.   0.   1.5  1.   0.  ]
 [1.   0.   3.2  1.   0.  ]
 [0.   0.   2.15 1.   0.  ]
 [1.   1.   4.2  0.   1.  ]
 [1.   0.   6.5  0.   1.  ]
 [1.   0.   0.5  0. 

Se convierten los datos a un ``dataframe`` de ``pandas``

In [ ]:
import pandas as pd

In [ ]:
# Crear el dataframe con estos datos, dando nombre a las columnas
delivery_data = pd.DataFrame(data, columns=["bad_weather", "is_rush_hour", "mile_distance", "urban_address", "late"])

In [ ]:
# Imprimir las primeras 5 filas
delivery_data.head(5)

bad_weather  is_rush_hour  mile_distance  urban_address  late
0          0.0           1.0            5.0            1.0   0.0
1          1.0           0.0            7.0            0.0   1.0
2          0.0           1.0            2.0            1.0   0.0
3          1.0           1.0            4.2            1.0   0.0
4          0.0           0.0            7.8            0.0   1.0

## 2. <a name="2">Selección de características para entrenar el modelo</a>

Vamos a separar las características de la variable objetivo

In [ ]:
input_data = delivery_data[["bad_weather", "is_rush_hour", "mile_distance", "urban_address"]]
target = delivery_data["late"]

En este conjunto de datos, que contiene muestras de cada una de las dos posibles clases, ajustamos un estimador de la biblioteca ``sklearn`` para capturar la relación entre la entrada y la salida, y exploramos más a fondo esa relación aprendida para predecir las clases a las que pertenecen las muestras no vistas.

En ``sklearn``, un estimador es un objeto de Python que implementa los métodos ``.fit()`` y ``.predict()``. El constructor del estimador toma como argumentos los parámetros del modelo.

## 3. <a name="2">Entrenamiento del clasificador</a>

Vamosw a ajustar a nuestros datos un clasificador de $k$ vecinos más cercanos. Utilizamos la librería de ``sklearn`` ``KNeighborsClassifier``.

![train-val-test split](https://drive.google.com/uc?id=1kybB4heQ_PqaDWddiKvU8K5EGioof5sv)


In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
# Use n_neighbors = 1
# Esto significa que el clasificador utilizará la etiqueta del vecino más cercano para hacer la predicción.
classifier = KNeighborsClassifier(n_neighbors = 1)

# Ajuste el modelo a los datos
classifier.fit(input_data, target)

KNeighborsClassifier(n_neighbors=1)

## 4. <a name="4">Uso del clasificador entrenado para predecir</a>
Ahora, vamos a predecir las etiquetas de nuevas muestras con nuestro clasificador entrenado.

In [ ]:
some_data = np.array([[0, 0, 2.1, 1]]) # bad_weather->0, is_rush_hour->0, mile_distance->2.1 and urban_address->1

# Utilice el modelo ajustado para hacer la predicción
print(classifier.predict(some_data))

[0.]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


Según nuestro modelo, este envío llegará a tiempo

## 5. <a name="5">Evaluación del modelo</a>

__¿Cómo sabemos si nuestro modelo hace predicciones buenas o malas?__

Si no tenemos la etiqueta real de este ejemplo (en este caso este pedido), no lo sabremos. Igualmente, no tenemos forma de saber qué tan bueno es este modelo.

Lo que podemos hacer es validar el modelo con los datos en los que lo entrenamos, y utilizar algunas métricas para examinar el rendimiento del clasificador.

In [ ]:
# Utilizar el modelo entrenado para predecir las etiquetas de nuestro conjunto de entrenamiento
predictions = classifier.predict(input_data)

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
print(confusion_matrix(target, predictions))

[[27  0]
 [ 0 18]]


Cuando observamos la matriz de confusión, podemos ver rápidamente que todas las predicciones fueron correctas, por lo que las métricas de clasificación deberían ser altas.

__Métricas de clasificación:__ Aquí usamos la métrica __tasa de acierto__ (accuracy), que mide cuán correctamente el modelo entrenado predice los resultados de retraso o no retraso. Veamos el informe de clasificación y la puntuación de accuracy a continuación.

In [ ]:
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

In [ ]:
print(classification_report(target, predictions))
print("Accuracy:", accuracy_score(target, predictions))

              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00        27
         1.0       1.00      1.00      1.00        18

    accuracy                           1.00        45
   macro avg       1.00      1.00      1.00        45
weighted avg       1.00      1.00      1.00        45

Accuracy: 1.0


## 6. <a name="6">Conjuntos de entrenamiento y prueba</a>

El modelo funcionó con un 100% de precisión en todo el conjunto de datos. Esto puede parecer prometedor, pero no nos dice nada sobre el rendimiento en futuros pedidos. Una forma de probar si este modelo funciona en nuevos pedidos "no vistos", es reservar algunos datos de nuestro conjunto de datos original para fines de prueba.

Dividamos nuestros datos en dos conjuntos: Entrenamiento (85%) y Prueba (15%). Esto nos dará 38 registros de entrenamiento y 7 registros de prueba (de los 45 registros totales).

![train-val-test split](https://drive.google.com/uc?id=18yLEOEuKylYPC_HZUagedK-9qYhwZzkK)

In [ ]:
delivery_data.shape

(45, 5)

In [ ]:
# Dividamos nuestra base de datos: Training (85%) and Test (15%)
training_data = delivery_data.iloc[:38, :] # First 38
test_data = delivery_data.iloc[38:, :] # Remaining

training_data.head()

bad_weather  is_rush_hour  mile_distance  urban_address  late
0          0.0           1.0            5.0            1.0   0.0
1          1.0           0.0            7.0            0.0   1.0
2          0.0           1.0            2.0            1.0   0.0
3          1.0           1.0            4.2            1.0   0.0
4          0.0           0.0            7.8            0.0   1.0

Ajustemos el $k$NN en los datos de entrenamiento esta vez:

In [ ]:
X_train = training_data[["bad_weather", "is_rush_hour", "mile_distance", "urban_address"]].values
y_train = training_data["late"].tolist()

# Use n_neighbors = 1
classifier = KNeighborsClassifier(n_neighbors = 1)

classifier.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=1)

In [ ]:
# Usemos el modelo ajustado para hacer predicciones en el conjunto de entrenamiento
train_predictions = classifier.predict(X_train)

print('Model evaluation on the training set: \n')
print(confusion_matrix(y_train, train_predictions))
print(classification_report(y_train, train_predictions))
print("Training accuracy:", accuracy_score(y_train, train_predictions))

Model evaluation on the training set: 

[[23  0]
 [ 0 15]]
              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00        23
         1.0       1.00      1.00      1.00        15

    accuracy                           1.00        38
   macro avg       1.00      1.00      1.00        38
weighted avg       1.00      1.00      1.00        38

Training accuracy: 1.0


Ahora, comprobemos el rendimiento en los datos de test

In [ ]:
X_test = test_data[["bad_weather", "is_rush_hour", "mile_distance", "urban_address"]].values
y_test = test_data["late"].tolist()

# Usemos el modelo ajustado para predecir las etiquetas del conjunto de test
test_predictions = classifier.predict(X_test)

print('Model evaluation on the training set: \n')
print(confusion_matrix(y_test, test_predictions))
print(classification_report(y_test, test_predictions))
print("Training accuracy:", accuracy_score(y_test, test_predictions))


Model evaluation on the training set: 

[[3 1]
 [1 2]]
              precision    recall  f1-score   support

         0.0       0.75      0.75      0.75         4
         1.0       0.67      0.67      0.67         3

    accuracy                           0.71         7
   macro avg       0.71      0.71      0.71         7
weighted avg       0.71      0.71      0.71         7

Training accuracy: 0.7142857142857143


## 7. <a name="7">Sobreajuste</a>

¡Esto no se ve bien!

Solo logramos un 71% de precisión en datos que el modelo no ha visto antes.
¿Podemos confiar en este modelo? Probablemente no.
Expliquemos lo que pasó aquí.

Experimentamos un problema común llamado "sobreajuste". Esto significa que nuestro modelo "aprendió demasiado" o memorizó nuestros datos de entrenamiento, y falló en los nuevos datos que no ha visto antes.

Las personas experimentadas habrían detectado el problema incluso antes de ajustar el clasificador, el parámetro $k$ que elegimos como 1 aquí mira el registro más cercano y asigna la clase de ese registro. Esto no se generaliza bien a nuestro conjunto de datos general y "sobreajusta" el conjunto de datos.
¿Dónde está el subconjunto de validación?

Si queremos optimizar el rendimiento de nuestro algoritmo, y por lo tanto reducir la llamada brecha de generalización, necesitamos buscar el valor de $k$ que mejor rendimiento tenga, utilizando un conjunto de validación. Elegimos el valor de $k$ que resulta en la mejor métrica de rendimiento de validación de nuestra elección, y luego finalmente comprobamos el rendimiento del modelo en el conjunto de datos de prueba.

In [ ]:
# Dividamos los datos de entremiento en dos subconjuntos: Training (80%) and Validation (20%)

train_data = training_data.iloc[:30, :] # First 30
val_data = training_data.iloc[30:, :] # Remaining

X_train = train_data[["bad_weather", "is_rush_hour", "mile_distance", "urban_address"]].values
y_train = train_data["late"].tolist()

X_val = val_data[["bad_weather", "is_rush_hour", "mile_distance", "urban_address"]].values
y_val =val_data["late"].tolist()


Entrenemos para diferentes valores de $k$

In [ ]:
K_values = [1, 2, 3, 4, 5, 6]

for K in K_values:
    classifier = KNeighborsClassifier(n_neighbors = K)
    classifier.fit(X_train, y_train)
    val_predictions = classifier.predict(X_val)
    print("K=%d, Validation accuracy score: %f" % (K, accuracy_score(y_val, val_predictions)))

K=1, Validation accuracy score: 0.875000
K=2, Validation accuracy score: 0.875000
K=3, Validation accuracy score: 1.000000
K=4, Validation accuracy score: 1.000000
K=5, Validation accuracy score: 1.000000
K=6, Validation accuracy score: 0.875000


Parece que $k = 3$ o $k = 4$ o $k = 5$ son opciones óptimas para $k$. Escojamos $k = 4$ para construir el clasificador, entrenar en el conjunto de entrenamiento y finalmente probar en el conjunto de prueba.

In [ ]:
classifier = KNeighborsClassifier(n_neighbors = 4)
classifier.fit(X_train, y_train)
test_predictions = classifier.predict(X_test)
print("Test accuracy score: %f" % (accuracy_score(y_test, test_predictions)))

Test accuracy score: 0.857143


En efecto, la precisión en el conjunto de prueba mejoró del 71% al 86%, reduciendo la brecha de generalización.